velox/exec/Exchange.cpp

```diff
@@ -58,127 +60,110 @@ void SerializedPage::prepareStreamForDeserialize(ByteStream* input) {
   if (strncmp(taskId.c_str(), "local://", 8) == 0) {
     return std::make_unique<LocalExchangeSource>(
-        taskId, destination, std::move(queue));
+        taskId, destination, std::move(queue), nullptr);
   }
   return nullptr;
 }
 
-} // namespace
+void LocalExchangeSource::request() {
+  auto buffers = PartitionedOutputBufferManager::getInstance().lock();
+  VELOX_CHECK_NOT_NULL(buffers, "invalid PartitionedOutputBufferManager");
+  VELOX_CHECK(requestPending_);
+  auto requestedSequence = sequence_;
+  auto self = shared_from_this();
+  buffers->getData(
+      taskId_,
+      destination_,
+      kMaxBytes,
+      sequence_,
+      // Since this lambda may outlive 'this', we need to capture a
+      // shared_ptr to the current object (self).
+      [self, requestedSequence, buffers, this](
+          std::vector<std::unique_ptr<folly::IOBuf>> data, int64_t sequence) {
+        if (requestedSequence > sequence) {
+          VLOG(2) << "Receives earlier sequence than requested: task "
+                  << taskId_ << ", destination " << destination_
+                  << ", requested " << sequence << ", received "
+                  << requestedSequence;
+          int64_t nExtra = requestedSequence - sequence;
+          VELOX_CHECK(nExtra < data.size());
+          data.erase(data.begin(), data.begin() + nExtra);
+          sequence = requestedSequence;
+        }
+        std::vector<std::unique_ptr<SerializedPage>> pages;
+        bool atEnd = false;
+        for (auto& inputPage : data) {
+          if (!inputPage) {
+            atEnd = true;
+            // Keep looping, there could be extra end markers.
+            continue;
+          }
+          inputPage->unshare();
+          pages.push_back(
+              std::make_unique<SerializedPage>(std::move(inputPage), pool_));
+          inputPage = nullptr;
+        }
+        int64_t ackSequence;
+        {
+          std::lock_guard<std::mutex> l(queue_->mutex());
+          requestPending_ = false;
+          for (auto& page : pages) {
+            queue_->enqueue(std::move(page));
+          }
+          if (atEnd) {
+            queue_->enqueue(nullptr);
+            atEnd_ = true;
+          }
+          ackSequence = sequence_ = sequence + pages.size();
+        }
+        // Outside of queue mutex.
+        if (atEnd_) {
+          buffers->deleteResults(taskId_, destination_);
+        } else {
+          buffers->acknowledge(taskId_, destination_, ackSequence);
+        }
+      });
+}
+
+bool LocalExchangeSource::shouldRequestLocked() {
+  if (atEnd_) {
+    return false;
+  }
+  bool pending = requestPending_;
+  requestPending_ = true;
+  return !pending;
+}
+
+void LocalExchangeSource::close() {
+  auto buffers = PartitionedOutputBufferManager::getInstance().lock();
+  buffers->deleteResults(taskId_, destination_);
+}
 
-void ExchangeClient::maybeSetMemoryPool(memory::MemoryPool* pool) {
+void ExchangeClient::initialize(memory::MemoryPool* pool) {
```

Comment:
The comment is stale now. I am wondering if we need to make the change here in this PR?

Reply:
The refactoring will also be appearing in another PR. After rebasing the changes here will disappear. I'm planning to land the other refactoring PR before this one.